### ISSUE TITLE
support other target machines (Atari, Vic20, BBC Micro...)

### ISSUE THREAD

**Issue by irmen** [irmen/prog8#6]
I'd like prog8 to better support other 6502 machines as well. But I know very little of the others, and any help here would be very much appreciated

Here is a guide to help fill out the blanks required to add a new target to the compiler: https://prog8.readthedocs.io/en/latest/portingguide.html

Here is some information about the customizable compiler target feature https://prog8.readthedocs.io/en/latest/targetsystem.html#customizable-target

**Comment by irmen** [comment]
Some progress has been made here recently, by adding a switch to select the compiler output target (and adding a preliminary c-language code generator)

**Comment by irmen** [comment]
The commanderX16 target is now the second functional target and it kinda ironed out the kinks that should make it fairly easy to add another target.  (as long as we know the machine's memory map and such)

**Comment by zbyti** [comment]
Are you considering at least basic support for 8-bit Atari?

**Comment by irmen** [comment]
I would love to!!   
But I don't know how they work.

What needs to be done is mainly implement the ICompilationTarget and IMachineDefinition objects for it that tell the compiler about memory layout and stuff.

**Comment by zbyti** [comment]
Safe memory location for code is from $2000 to $BFFF when BASIC is off - this default scenario.

Most simple binary (XEX) for Atari is is build this way:

`$FF $FF start-LO begining-HI begining-LO end-HI [ bytes of our program ] start-LO start-HI`

https://www.atarimax.com/jindroush.atari.org/afmtexe.html

There are other headers for multi segments programs but above is good for start :]

---

```
2000: 4C 03 20  JMP $2003
2003: AD 0B D4  LDA $D40B   ;VCOUNT
2006: 8D 18 D0  STA $D018   ;COLPF2
2009: 8D 0A D4  STA $D40A   ;WSYNC
200C: 4C 03 20  JMP $2003
```

as XEX file looks this way:

`FF FF 00 20 0E 20 4C 03 20 AD 0B D4 8D 18 D0 8D 0A D4 4C 03 20 E0 02 E1 02 00 20`

Header: first 6 bytes and last two.

This way you can load one big segment into memory from $2000 to $200E and run program from $2000.

**Comment by zbyti** [comment]
Oh, well I forgot about:

`E0 02 E1 02`

:D

**Comment by irmen** [comment]
What type of 8 bit atari are we talking about anyway?
Also can you recommend a good emulator for it (preferably running on Linux)?

**Comment by zbyti** [comment]
Atari 800 XL 64KB RAM.

On Ubuntu I use https://atari800.github.io/ - in special cases I use Altirra under WINE.

Run atarti800 from command-line then F8 give you acces to the monitor.

**Comment by zbyti** [comment]
Atari 800XL code examples:

https://github.com/zbyti/k65-playground/tree/main/src/a8

https://github.com/tebe6502/Mad-Pascal/tree/master/samples/a8

**Comment by irmen** [comment]
how to print some text to the screen on 800xl?

**Comment by zbyti** [comment]
With OS: `COLCRS ($55-$56) i ROWCRS ($54).`

Without OS you must take screen memory pointer from `SAVMSC` located at  `$58`, `$59`.

https://www.atariarchives.org/mapping/memorymap.php

**Comment by zbyti** [comment]
I'm not an expert, I always turn off system and write my own procedures to write on screen ;)

**Comment by irmen** [comment]
well at least 64tass supports generating xex files so we don't have to adapt the tool chain.

**Comment by irmen** [comment]
I don't like the idea of writing my own PRINT routines, I much rather use a rom routine to put text and numbers on the screen. Is there a list of atari rom routines like this somewhere?

**Comment by zbyti** [comment]
something like this? https://github.com/ilmenit/A800-OS-XL-Rev2

**Comment by irmen** [comment]
@zbyti to streamline things in a way to get going, I really need at least parts of a questionnaire filled out for me. Here is the (new) porting guide I wrote: Here is a guide to help fill out the blanks required to add a new target to the compiler: https://prog8.readthedocs.io/en/latest/portingguide.html

As I know next to nothing about the Atari 8bits, I really need someone (you?) to step in and fill in those questions to get things on the road.   Not all are required to be answered to get at least a bare bones thing running, but some really are, and I don't have the time to investigate this platform myself now.

**Comment by zbyti** [comment]
Okay, I'll try to answer these questions soon, I'll consulting with someone more experienced on this platform than me if necessary.

A8 support is not urgent, you know by your self how many people programing on 8-bit in anything other than ASM or BASIC ;)

**Comment by irmen** [comment]
C128 target was added in 7.6 (but very incomplete still).

**Comment by zbyti** [comment]
Atari XL support: @FreddyOffenga explains everything here: https://www.youtube.com/watch?v=ezQuRA1FPJ0 ;)

**Comment by FreddyOffenga** [comment]
@zbyti thanks for refering to my presentation :)

You can also find the textual info here:
https://github.com/FreddyOffenga/sizecoding/blob/main/presentations/Atari8bit/presentationA8.md

JAC! did a much longer tutorial about the Atari 8-bit which might be useful if you want to know the basics of the platform. Part 3 and 4 are about the memory map.
https://www.wudsn.com/index.php/productions-atari800/tutorials

Keep up the good work!

**Comment by irmen** [comment]
@FreddyOffenga @zbyti I'd really like to add atari XL support to prog8 but I don't have the time to comb through the materials myself right now. 
If you could kickstart it for me by answering the questions in the porting guide (https://prog8.readthedocs.io/en/latest/portingguide.html) it makes things a lot easier for me. Then, once an initial atari XL target is added based on that info , you can start experimenting with it and take it from there.

**Comment by zbyti** [comment]
# CPU
1. 6502
2. yes

# Memory Map
## Zero page

Programs not written in BASIC do not require BASIC to run. On Atari, programs run under DOS (many different variants).

1.  $00CB-$00CF unused by BASIC and ASSEMBLER
2. https://atariwiki.org/wiki/Wiki.jsp?page=Memory%20Map
3. $80-$ff without $00D4-$00E5
4. $80-$ff

## RAM, ROM, I/O

1. depends on used DOS, most common is $2000-$bfff but ANTIC reserve some memory, depending on the used gfx mode, from RAMTOP to down.
2. not applicable
3-4. $0600-$bfff if you have in mind DOS, BASIC & ANTIC
5. $a000-$bfff if basic on, generally without HW registers on $dxxx ROM covers RAM from $c000 up.
6. $dxxx look at: https://atariwiki.org/wiki/Wiki.jsp?page=Memory%20Map
7. don't know in details, not used yet.

From here I count on @FreddyOffenga  ;)

**Comment by irmen** [comment]
master branch now has an ``atari`` target, it creates a .xex file but it doesn't run yet because the details need a lot of work still. Maybe you can have a look at it already though to see if this is going anywhere?

**Comment by FreddyOffenga** [comment]
Cool thank you! Will take a look and see where or how I can help, probably this weekend.

**Comment by ari26squared** [comment]
Does Prog8 support custom targets?

**Comment by irmen** [comment]
what do you mean with that?

**Comment by ari26squared** [comment]
I mean - can i target a 65c02 with a custom memory map and io? for examples /r/beneater

**Comment by irmen** [comment]
Ah I see. Well, no it doesn't - the target machine's configurations are hardcoded right now. It's not much work to add a new configuration though if you know what's needed.

**Comment by ari26squared** [comment]
oh i see. im not really all that familiar with what would be needed and also the languages used. im basically looking for something akin to cc65 customization but with a better C standard.

**Comment by irmen** [comment]
Well prog8 is not C, so if you're looking for "a better C standard" you're looking in the wrong place.

**Comment by ari26squared** [comment]
alternatives to C89 is preferred

**Comment by irmen** [comment]
@RealDoigt  that is not an easy thing to answer.

But let's start by saying that many of the questions in the "porting guide" linked in the original message, https://prog8.readthedocs.io/en/latest/portingguide.html , have to be answered.

This knowledge is then encoded into the compiler and support libraries.

Most notably, in the various target-specific classes inside the CodeCore module, look here  https://github.com/irmen/prog8/tree/master/codeCore/src/prog8/code/target

And also most importantly in the syslib library module, look here for the Commander X16 version https://github.com/irmen/prog8/blob/master/compiler/res/prog8lib/cx16/syslib.p8 

It's been a while since I added a new target, but the rest of the compiler should be (near) fully agnostic of the target platform. As long as it has a 6502 type cpu

**Comment by RealDoigt** [comment]
Yes, that is unfortunate but you are right, not being able to store programs in a rom is a major deal breaker right out of the door.

**Comment by irmen** [comment]
Maybe you can put RAM on the nes cartridge? 
Anyway, for the time being I've removed NES from the list of probable additional target systems...

**Comment by irmen** [comment]
There's a `neo6502` branch containing initial support for the Neo6502 board https://neo6502.com/

**Comment by irmen** [comment]
The 'neo' branch has just been merged into the master branch, so it's available now by default (though very bare bones still)

**Comment by y0014984** [comment]
I would be happy if a very simple target system config file would be supported instead of hard-coded targets. Currently, I'm working on my own 6502-based system and the memory map and variable locations are changing quite often. I would like to test Prog8 for my development but choosing a non matching target seems to be the wrong way for me. From my understanding there isn't more needed than defining memory regions (in zeropage and elsewhere) that should be ignored by Prog8, a start address for the program and the CPU type. I whould be happy if you could add such a feature. If it's more sophisticated you could distinguish between ROM, Used RAM, Unused RAM, Used ZP, Unused ZP.

**Comment by irmen** [comment]
I've started an experiment to have a .properties file that can be specified as a compilation target, in which you can then define the variables. Something like this:

```properties
# configuration file for a flexible Prog8 compilation target

cpu = 65C02
program = CBMPRG
encoding = iso
loadaddress = $0801
memtop = $d000
startup_reserved = 32
bsshighram_start = 0
bsshighram_end = 0
bssgoldenram_start = 0
bssgoldenram_end = 0

# ioaddresses can be zero or more memory address ranges (inclusive) separated by comma
ioaddresses = $d000-$dfff

# zeropage scratch variables. reg must always be b1+1 !
zp_scratch_b1 = $e0
zp_scratch_reg = $e1
zp_scratch_w1 = $e2
zp_scratch_w2 = $e4

# the start of the 32 bytes used by the R0-R15 virtual registers. Can be in Zeropage or elsewhere.
virtualregisters = $10

# TODO some more things still need to follow...
```

**Comment by irmen** [comment]
Btw, I'll probably announce and discuss the configurability feature on the discord channel more directly, rather than spamming here.

**Comment by RealDoigt** [comment]
We're not all on your discord

**Comment by irmen** [comment]
Then you have to wait and see what it will or will not end up as

**Comment by RealDoigt** [comment]
as long as the information is not private, I'm fine with that

**Comment by irmen** [comment]
Sure, as always, it will be documented in the manual

**Comment by irmen** [comment]
The very bare-bones atari and neo6502 targets have been moved out of the compiler to externally configurable targets
hopefully this makes them easier to adapt and improve (by others?)

**Comment by adiee5** [comment]
@RealDoigt btw, idk if you've already seen it (It's technically mentioned in `examples/customtargets`), but there's now an [experimental NES target](https://github.com/adiee5/prog8-nes-target)

**Comment by RealDoigt** [comment]
Thank you for bringing this to my attention, I'll take a look at it tonight and I'll see where I can help. I think I could probably help with sound once the hp stuff is done.

**Comment by country-bumpkin-software** [comment]
> Is anyone working on targeting the MEGA65 or the Foenix F256K?

it would be great if it were available to take advantage of the mega65

**Comment by irmen** [comment]
Indeed i hope someone familiar with those systems will write a custom target configuration.  It shouldn't be too hard anymore to do that now that almost everything system-specific is configurable in the custom target config file

**Comment by adiee5** [comment]
Thanks for thinking about it. I didn't work yet on the target this summer either.

**Comment by irmen** [comment]
More custom prog8 targets (such as VIC-20) available here https://github.com/gillham/prog8targets

**Comment by irmen** [comment]
closing this issue as I think the custom targets feature is now more than enough to build more compilation targets
